lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -3182,6 +3182,20 @@ LogicalResult InvalidValueOp::canonicalize(InvalidValueOp op,
     rewriter.eraseOp(op);
     return success();
   }
+  // Propagate invalids through a single use which is a unary op.  You cannot
+  // propagate through multiple uses as that breaks invalid semantics.  Nor
+  // can you propagate through binary ops.
+  if (op->hasOneUse() &&
+      isa<BitsPrimOp, HeadPrimOp, ShrPrimOp, TailPrimOp, SubfieldOp, SubindexOp,
+          AsSIntPrimOp, AsUIntPrimOp, NotPrimOp>(*op->user_begin())) {
```

Comment:
There are a few more unwary ops that can be included: cvt, pad, asClock, and the reduction ops.

Reply:
Pad adds bits, so it isn't safe, cvt interprets the bits and can add bits, so likewise.  asClock is probably safe.

Reply:
But reductions are a good call.

Reply:
oddly, a reduction of a zero-width value is NOT invalid, even though all other widths are.